apm-agent-core/src/main/java/co/elastic/apm/agent/util/JvmRuntimeInfo.java

```diff
@@ -0,0 +1,201 @@
+            }
+            try {
+                updateVersion = Integer.parseInt(updateVersionString);
+            } catch (NumberFormatException e) {
+                // in case of unknown format, we just support by default
+                System.err.println("Unsupported format of the java.version system property - " + version);
+                updateVersion = -1;
+            }
+        }
+
+        isHotSpot = vmName.contains("HotSpot(TM)") || vmName.contains("OpenJDK");
+        isIbmJ9 = vmName.contains("IBM J9");
+    }
+
+    public static String getJavaVersion() {
+        return javaVersion;
+    }
+
+    public static String getJavaVmName() {
+        return javaVmName;
+    }
+
+    @Nullable
+    public static String getJavaVmVersion() {
+        return javaVmVersion;
+    }
+
+    public static int getMajorVersion() {
+        return majorVersion;
+    }
+
+    /**
+     * Checks if a given version of the JVM is likely supported by this agent.
+     * <br>
+     * Supports values provided before and after https://openjdk.java.net/jeps/223, in case parsing fails due to an
+     * unknown version format, we assume it's supported, thus this method might return false positives, but never false
+     * negatives.
+     *
+     * @return true if the version is supported, false otherwise
+     */
+    public static boolean isJavaVersionSupported() {
+        if (majorVersion < 7) {
+            // given code is compiled with java 7, this one is unlikely in practice
+            return false;
+        }
+        if (isHotSpot) {
+            return isHotSpotVersionSupported();
+        } else if (isIbmJ9) {
+            return isIbmJ9VersionSupported();
+        }
+        // innocent until proven guilty
+        return true;
+    }
+
+    private static boolean isHotSpotVersionSupported() {
+        if (updateVersion < 0) {
+            return true;
+        }
+
+        switch (majorVersion) {
+            case 7:
+                // versions prior to that have unreliable invoke dynamic support according to https://groovy-lang.org/indy.html
+                return updateVersion >= 60;
+            case 8:
+                return updateVersion >= 40;
+            default:
+                return true;
+        }
+    }
+
+    private static boolean isIbmJ9VersionSupported() {
+        switch (majorVersion) {
+            case 7:
+                return false;
+            case 8:
+                // early versions crash during invokedynamic bootstrap
+                // the exact version that fixes that error is currently not known
+                // presumably, service refresh 5 (build 2.8) fixes the issue
+                return !"2.8".equals(javaVmVersion);
+            default:
+                return true;
+        }
+    }
+
+    /**
+     * NOTE: THIS METHOD REQUIRES JMX OPERATION, THEREFORE IS SHOULD ONLY BE USED WITHIN OR AFTER THE TRACER HAVE
```

Comment:
Could we check for tracer state and throw an exception if triggered before tracer has been started ?

Reply:
I didn't want to do any JMX query at initialization time because of my JULI deadlock trauma (this kicks all kind of initializations). 
However, @felixbarny feels it is safe enough to do this simple JMX query very early on, as long as we don't try to `getPlatformMBeanSerever`. It certainly simplifies things a lot (no need to uninstrument classes just after they are done being instrumented). So we'll go with that. I will test and merge the change soon

Reply:
I'd also recommend adding a way to disable the bootstrap checks, for example with a system property, in case it does cause problems. It can't be a config option as these get initialized later.